### PR TITLE
OOP

### PR DESCRIPTION
**Pull request by natashafir**
### Frogger game

[Demo](https://natashafir.github.io/frogger-game/) |
[Code base](https://github.com/natashafir/frogger-game/blob/main/js/app.js)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@natashafir nice job!

It seems like bugs at their first appearance storm through the scene like jets! Not sure if it was intentional.

There are few flaws in the code to fix.

A checklist to your rescue.

 * [ ] it is OK to employ ES6 features like `const`, `let` etc.
 * [ ] the code is very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
 * [ ] OO is implemented using JS prototype chain object model (**not** ES6 classes syntax)
 * Requirements re **Constants**:
   * [ ] all numbers like block dimensions, initial locations are defined as named constants
         (e.g. `const STEP = 101;`) as otherwise numbers scattered across code base look cryptic;
         named constants add semantic meaning and improve readability
   * [ ] there are core constants and derived constants
         (e.g. derived constant `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
   * [ ] arrays of constants are also constants 
         (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
   * [ ] const objects help organizing and structure const data even better
         (e.g. `const PLAYER_CONF = { initialPosition: {x: 1, y: 5}, sprite: '...', ...etc... };`
 * Requirements re **OOP**:
   * [ ] classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
         (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
         Hint: pass `Player` instance as an argument to every enemy
   * [ ] Separation of Concerns principle is followed
         (e.g. `update` method does only rendering and doesn't contain any unrelated **inline** code;
         for example collision check is defined as a dedicated method and only called from inside `update`)
   * [ ] Nice To Have: properties common for some classes are generalized into a base class
         (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
   * [ ] class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
         not `Subclass.prototype = new Superclass(params);`;
         [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
 * Most common mistakes
   * [ ] Make sure `target = condition ? valueWhenConditionTrue : valueWhenConditionFalse` is used instead of
         `condition ? target = valueWhenConditionTrue : target = valueWhenConditionFalse`;
         [Conditional (ternary) operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)

Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

**Comment by natashafir** [comment]
@OleksiyRudenko thank you a lot for the review, I will make an effort to improve my code.

File: submissions/natashafir/oop/app.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+            player.x = INITIAL_POSITION_X;
+            player.y = INITIAL_POSITION_Y;
+        }, 200)
+    }
+};
+
+let bugLocation = [60, 145, 230];
+
+bugLocation.forEach(function (bugY) {
+    let enemy = new Enemy(BUGS_START, bugY, 200);
```
Comment (OleksiyRudenko) [inline submissions/natashafir/oop/app.js:94]: There must be a semantic meaning behind mysterious number `200`. Tell a reader what it is without browsing the codebase.

File: submissions/natashafir/oop/app.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    if (this.y < 0) {
+        setTimeout(function () {
+            player.x = INITIAL_POSITION_X;
+            player.y = INITIAL_POSITION_Y;
+        }, 200)
+    }
+};
+
+let bugLocation = [60, 145, 230];
```
Comment (OleksiyRudenko) [inline submissions/natashafir/oop/app.js:91]: Please demistify these numbers.
Can these be calculated from some basic values? Are these some pixel positions calculable from the game scene row number plus or minus certain offset?

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@natashafir a really good job!
A few fixes and we're done

File: submissions/natashafir/oop/app.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+// You should multiply any movement by the dt parameter
+// which will ensure the game runs at the same speed for
+// all computers.
+Enemy.prototype.update = function (dt) {
+    this.x += this.speed * dt;
+    if (this.x > BUGS_END) {
+        this.x = BUGS_START;
+        this.speed = (Math.floor(Math.random() * MIN_SPEED) + MAX_SPEED) * dt;
+    }
+    if (this.Collision()) player.resetPosition();
```
[the remark this answers is not in the document]
Comment (natashafir) [reply submissions/natashafir/oop/app.js:33]: should be ```this.player.resetPosition()```?

**Review by OleksiyRudenko — APPROVED** [review]
@natashafir great job!

**Pull request merged**